File: mobile/src/main/java/technology/mainthread/apps/gatekeeper/data/GatekeeperAuthManager.java
package technology.mainthread.apps.gatekeeper.data;

import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableOnSubscribe;
import technology.mainthread.apps.gatekeeper.common.SharedValues;
import timber.log.Timber;

public class GatekeeperAuthManager implements AuthManager {

    private final GoogleApiClient googleApiClient;
    private final FirebaseAuth firebaseAuth;
    private final RegisterDevices registerDevices;

    public GatekeeperAuthManager(GoogleApiClient googleApiClient,
                                 FirebaseAuth firebaseAuth,
                                 RegisterDevices registerDevices) {
        this.googleApiClient = googleApiClient;
        this.firebaseAuth = firebaseAuth;
        this.registerDevices = registerDevices;
    }

    @Override
    public GoogleApiClient getGoogleApiClient() {
        return googleApiClient;
    }

    @Override
    public Flowable<Boolean> authWithFirebase(GoogleSignInAccount account) {
        FlowableOnSubscribe<Boolean> source = e -> {
            e.onNext(authWithFirebaseSync(account));
            e.onComplete();
        };
        return Flowable.create(source, BackpressureStrategy.BUFFER);
    }

    @Override
    public Flowable<Boolean> signOut() {
        FlowableOnSubscribe<Boolean> source = e -> {
            e.onNext(signOutSync());
            e.onComplete();
        };
        return Flowable.create(source, BackpressureStrategy.BUFFER);
    }

    @Override
    public Flowable<Boolean> deleteAccount() {
        FlowableOnSubscribe<Boolean> source = e -> {
            e.onNext(deleteAccountSync());
            e.onComplete();
        };
        return Flowable.create(source, BackpressureStrategy.BUFFER);
    }

    private ConnectionResult connect() {
        return googleApiClient.blockingConnect(SharedValues.CONNECTION_TIME_OUT_MS, TimeUnit.MILLISECONDS);
    }

    private boolean authWithFirebaseSync(GoogleSignInAccount acct) {
        AuthCredential credential = GoogleAuthProvider.getCredential(acct.getIdToken(), null);
        Task<AuthResult> task = firebaseAuth.signInWithCredential(credential);
        try {
            Tasks.await(task);
        } catch (ExecutionException e) {
            Timber.d(e, "task await ExecutionException");
        } catch (InterruptedException e) {
            Timber.e(e, "task await InterruptedException");
        }

        boolean success = task.isComplete() && task.isSuccessful() && task.getResult().getUser() != null;

        if (success) {
            registerDevices.registerDevice();
        }

        return success;
    }

    private boolean signOutSync() {
        firebaseAuth.signOut();
        boolean result = connect().isSuccess() && Auth.GoogleSignInApi.signOut(googleApiClient).await().isSuccess();
        if (googleApiClient.isConnected()) {
            googleApiClient.disconnect();
        }
        return result;
    }

    private boolean deleteAccountSync() {
        FirebaseUser user = firebaseAuth.getCurrentUser();
        // TODO: need to handle re-authenticating - https://firebase.google.com/docs/auth/android/manage-users#delete_a_user
        boolean result = user.delete().isSuccessful() && connect().isSuccess() && Auth.GoogleSignInApi.revokeAccess(googleApiClient).await().isSuccess();
        if (googleApiClient.isConnected()) {
            googleApiClient.disconnect();
        }
        return result;
    }
}

File: mobile/src/main/java/technology/mainthread/apps/gatekeeper/data/GatekeeperAuthManager.kt
package technology.mainthread.apps.gatekeeper.data

import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.tasks.Tasks
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import io.reactivex.BackpressureStrategy
import io.reactivex.Emitter
import io.reactivex.Flowable
import technology.mainthread.apps.gatekeeper.common.CONNECTION_TIME_OUT_MS
import timber.log.Timber
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit

class GatekeeperAuthManager(override val googleApiClient: GoogleApiClient,
                            private val firebaseAuth: FirebaseAuth,
                            private val registerDevices: RegisterDevices) : AuthManager {

    override fun authWithFirebase(account: GoogleSignInAccount): Flowable<Boolean> {
        val source = { e: Emitter<Boolean> ->
            e.onNext(authWithFirebaseSync(account))
            e.onComplete()
        }
        return Flowable.create<Boolean>(source, BackpressureStrategy.BUFFER)
    }

    override fun signOut(): Flowable<Boolean> {
        val source = { e: Emitter<Boolean> ->
            e.onNext(signOutSync())
            e.onComplete()
        }
        return Flowable.create<Boolean>(source, BackpressureStrategy.BUFFER)
    }

    override fun deleteAccount(): Flowable<Boolean> {
        val source = { e: Emitter<Boolean> ->
            e.onNext(deleteAccountSync())
            e.onComplete()
        }
        return Flowable.create<Boolean>(source, BackpressureStrategy.BUFFER)
    }

    private fun connect(): ConnectionResult {
        return googleApiClient.blockingConnect(CONNECTION_TIME_OUT_MS, TimeUnit.MILLISECONDS)
    }

    private fun authWithFirebaseSync(acct: GoogleSignInAccount): Boolean {
        val credential = GoogleAuthProvider.getCredential(acct.idToken, null)
        val task = firebaseAuth.signInWithCredential(credential)
        try {
            Tasks.await(task)
        } catch (e: ExecutionException) {
            Timber.d(e, "task await ExecutionException")
        } catch (e: InterruptedException) {
            Timber.e(e, "task await InterruptedException")
        }

        val success = task.isComplete && task.isSuccessful && task.result.user != null

        if (success) {
            registerDevices.registerDevice()
        }

        return success
    }

    private fun signOutSync(): Boolean {
        firebaseAuth.signOut()
        val result = connect().isSuccess && Auth.GoogleSignInApi.signOut(googleApiClient).await().isSuccess
        if (googleApiClient.isConnected) {
            googleApiClient.disconnect()
        }
        return result
    }

    private fun deleteAccountSync(): Boolean {
        val user = firebaseAuth.currentUser
        // TODO: need to handle re-authenticating - https://firebase.google.com/docs/auth/android/manage-users#delete_a_user
        val result = user!!.delete().isSuccessful && connect().isSuccess && Auth.GoogleSignInApi.revokeAccess(googleApiClient).await().isSuccess
        if (googleApiClient.isConnected) {
            googleApiClient.disconnect()
        }
        return result
    }
}
